### PR TITLE
New refactored video app

### PR DESCRIPTION
**Pull request by deepakrawat22**
This new refactored app is same as previous one in functionality, just implemented in a clean way.

GearVRf-DCO-1.0-Signed-off-by: 
Deepak Rawat deepak.rawat@samsung.com

### CODE REVIEW

File: gvr-video/app/src/main/assets/gvr.xml [hunk 1]
```diff
@@ -39,7 +39,7 @@
             colorFormat="COLOR_8888"
             depthFormat="DEPTH_24"
             fov-y="90.0"
-            multiSamples="1"
+            multiSamples="4"
```
Comment (liaxim) [inline gvr-video/app/src/main/assets/gvr.xml]: Why 4 and not 2?

Comment (deepakrawat22) [inline gvr-video/app/src/main/assets/gvr.xml]: With 2 the seekbar was very blurred, with 4 it looks smooth.

**Comment by thomasflynn** [comment]
as long as performance is still 60Hz on both mali and adreno, upping it to 4 should be ok.

**Comment by deepakrawat22** [comment]
The performance is almost same if set multisampe to 2 or 4, but I observed that its not always 60Hz, sometimes below 60 also.

**Comment by liaxim** [comment]
Have you tried Note4?

**Comment by deepakrawat22** [comment]
@liaxim Yes on Note 4 the FPS is also is around 60, occasionally it will go below 60.

**Comment by liaxim** [comment]
Needs a rebase.

**Comment by deepakrawat22** [comment]
Rebased to latest.

**Comment by NolaDonato** [comment]
I tried testing this with the latest GearVRf master. It builds OK but does not run. Here is the logcat:
 Executing tasks: [:app:clean, :app:generateDebugSources, :app:generateDebugAndroidTestSources, :app:mockableAndroidJar, :app:prepareDebugUnitTestDependencies, :app:assembleDebug]

Configuration on demand is an incubating feature.
 Incremental java compilation is an incubating feature.
 :app:clean
 :app:copyOculusSignature UP-TO-DATE
 :app:preBuild UP-TO-DATE
 :app:preDebugBuild UP-TO-DATE
 :app:checkDebugManifest
 :app:preReleaseBuild UP-TO-DATE
 :app:prepareFrameworkDebugLibrary
 :app:prepareGoogleVrBaseLibrary
 :app:prepareGoogleVrCommonLibrary
 :app:prepareDebugDependencies
 :app:compileDebugAidl
 :app:compileDebugRenderscript
 :app:generateDebugBuildConfig
 :app:mergeDebugShaders
 :app:compileDebugShaders
 :app:generateDebugAssets
 :app:mergeDebugAssets
 :app:generateDebugResValues UP-TO-DATE
 :app:generateDebugResources
 :app:mergeDebugResources
 :app:processDebugManifest
 :app:processDebugResources
 :app:generateDebugSources
 :app:preDebugAndroidTestBuild UP-TO-DATE
 :app:prepareDebugAndroidTestDependencies
 :app:compileDebugAndroidTestAidl
 :app:processDebugAndroidTestManifest
 :app:compileDebugAndroidTestRenderscript
 :app:generateDebugAndroidTestBuildConfig
 :app:mergeDebugAndroidTestShaders
 :app:compileDebugAndroidTestShaders
 :app:generateDebugAndroidTestAssets
 :app:mergeDebugAndroidTestAssets
 :app:generateDebugAndroidTestResValues UP-TO-DATE
 :app:generateDebugAndroidTestResources
 :app:mergeDebugAndroidTestResources
 :app:processDebugAndroidTestResources
 :app:generateDebugAndroidTestSources
 :app:mockableAndroidJar UP-TO-DATE
 :app:preDebugUnitTestBuild UP-TO-DATE
 :app:prepareDebugUnitTestDependencies
 :app:incrementalDebugJavaCompilationSafeguard
 :app:compileDebugJavaWithJavac
 :app:compileDebugJavaWithJavac - is not incremental (e.g. outputs have changed, no previous execution, etc.).

Note: Some input files use or override a deprecated API.
 Note: Recompile with -Xlint:deprecation for details.

:app:compileDebugNdk UP-TO-DATE
 :app:compileDebugSources
 :app:buildInfoDebugLoader
 :app:transformClassesWithExtractJarsForDebug
 :app:transformClassesWithInstantRunVerifierForDebug
 :app:transformClassesWithJavaResourcesVerifierForDebug UP-TO-DATE
 :app:mergeDebugJniLibFolders
 :app:transformNative_libsWithMergeJniLibsForDebug
 :app:processDebugJavaRes UP-TO-DATE
 :app:transformResourcesWithMergeJavaResForDebug
 :app:transformResourcesAndNative_libsWithJavaResourcesVerifierForDebug UP-TO-DATE
 :app:transformClassesWithInstantRunForDebug
 :app:transformClasses_enhancedWithInstant+reloadDexForDebug UP-TO-DATE
 :app:incrementalDebugTasks
 :app:prePackageMarkerForDebug
 :app:fastDeployDebugExtractor
 :app:generateDebugInstantRunAppInfo
 :app:transformClassesWithDexForDebug FAILED

FAILURE: Build failed with an exception.

•What went wrong:
 Execution failed for task ':app:transformClassesWithDexForDebug'.


com.android.build.api.transform.TransformException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.dex.DexException: Multiple dex files define Lcom/google/vr/cardboard/R$color;


•Try:
 Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.


BUILD FAILED

Total time: 18.824 secs

**Comment by liaxim** [comment]
It does build and run for me. @NolaDonato I believe there might be something wrong with your build environment.

**Comment by NolaDonato** [comment]
I built your app and tested it and took a look at the code. Here are my suggestions:

1. Use the new event-based picking approach rather than calling GVRPicker.pickScene every frame. GVRPicker.doPick does basically what FocusableController.processFocus does for you. 
2. Use the new method of creating custom shaders by subclassing GVRShaderTemplate. 

I tried to attach a sample of what I mean  (which I actually tested with your app) but I was not able to. I will email it to you.

**Comment by deepakrawat22** [comment]
Working on the suggestions given by @NolaDonato

**Comment by deepakrawat22** [comment]
This pull request is ready for review. Used new event-based picking approach. The new method of shader creation is not working, so not done this time. Also fixed some bugs.

**Comment by NolaDonato** [comment]
What was wrong with the new method of shader creation? I tested the code I gave you and it seemed to work OK on the Note 4 once we added support for using shader model 100.

**Comment by deepakrawat22** [comment]
I was not able to see any color in the cinema environment.

**Comment by NolaDonato** [comment]
I pulled and tested the video app. It works great. The thing is - the new method of shader creation should have worked. I verified what Deepak observed - when going thru the custom shader path something is not right with the textures. I will look into this because it has implications for other shader use as well.

**Comment by thomasflynn** [comment]
@NolaDonato is this good to go for now?

**Comment by liaxim** [comment]
If I start using the seekbar eventually the video playback freezes. The timer keeps ticking but the same frame is shown throughout. Happens always for me.

--

I suggest we re-implement the shaders in a subsequent pull request. This is a nice enough cleanup already.

**Comment by deepakrawat22** [comment]
@liaxim I was not able to reproduce this issue.

**Comment by liaxim** [comment]
Will show you on my phone.

**Comment by liaxim** [comment]
@deepakrawat22 reproduced. Looks like it might have something to do with the change I made to the videoSceneObject to support the ExoPlayer.

**Comment by liaxim** [comment]
https://github.com/Samsung/GearVRf/pull/858 fixes the reference to the video player so pause/resume works.

**Comment by deepakrawat22** [comment]
@liaxim verified now pause/resume works normally.

**Comment by liaxim** [comment]
Planning to merge in the afternoon if there are no objections. The shaders will be updated in a following pull request once the issues have been resolved.

**Comment by NolaDonato** [comment]
Yes merge this. I am still working on why the shader changes don't work.



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Pull request merged**